### ISSUE TITLE
Error code for inaccessible save path

### ISSUE THREAD

**Issue by sledgehammer999** [arvidn/libtorrent#240]
I want a new error code for this situation:
Adding a torrent **with a fastresume** and feeding a savepath (or letting the one in the fastresume). But that savepath doesn't exist. This could be happening because the user didn't connect his external drive or he forgot to mount the partition or due to permission issues. In this case the user expects to not lose progress and the torrent be paused or put in a special state.

Currently I am somewhat achieving this by looking for `errors::mismatching_file_size` in `fastresume_rejected_alert` alert. However, I feel that this might not cover all cases.

Is it possible to add this?

Furthermore, as an extra point, but not critical to this: Is it possible to specify (maybe in add_torrent_params) a way to automatically pause the torrent if the fastresume gets rejected?
Currently, I do the pause inside my `fastresume_rejected_alert` handler but the handler runs in parallel with the libtorrent thread. So in a rare case, libtorrent might go ahead and start downloading (and maybe overwrite files and progress) before I pause the torrent.

**Comment by arvidn** [comment]
I'm trying to think about what happens in this case today. I believe libtorrent is quite lazy in trying to write to the disk. It probably won't do that until it has downloaded a whole piece basically.

It will stat() every file though, and I would imagine that would fail with ENOENT. Can you think of anything to distinguish this case from the case where the destination files were just deleted?

I have a feeling that regardless of which indicator or heuristic we would choose would be unintuitive to users and have false positives and false negatives. How would you feel about adding another flag to add_torrent_params indicating that you expect the files (or at least the path) to be there?

another, possibly simpler, thing you could do would be to stat() the save path directory to see if it exists.

**Comment by sledgehammer999** [comment]
>Can you think of anything to distinguish this case from the case where the destination files were just deleted?

My simple test would be does "save path" exist? And do that test only when resuming. And you said it yourself:

>another, possibly simpler, thing you could do would be to stat() the save path directory to see if it exists.
    
About the stat() and relevant discussion, sorry but I cannot comment. I don't know that kind of stuff.

>How would you feel about adding another flag to add_torrent_params indicating that you expect the files (or at least the path) to be there?

That would be great. At least for the path. For the files, I guess it would be useful when you want to react to the "rejected fastresume" alert without having libtorrent overwrite them.
But I think it needs a specific error_code for this condition.

#### I know this is out of the original scope:
Would it be possible to have libtorrent retry verification of the fastresume data while the torrent is already in session? That will allow us to preserve queue position, if the user plugs in the drive (or mounts the partition) and wants to tell qbt "hey check again".

**Comment by arvidn** [comment]
checking again after having started is complicated (or to leave it in a state where it isn't started). the different states a torrent can be in and all events that can happen to it is already too large to be easily manageable.

My understanding is that on unix systems it's possible, if not likely, that a _path_ will exist even though the drive is not mounted. Perhaps it's limited to the root of the mount-point though, but still, it suggests that just checking to see if the path is there wouldn't be enough (because nothing can exist under that directory until the drive is mounted).

as for a new flag, I basically imagine that being an option to just stop the torrent if check_resume fails for any reason, or something along those lines

**Comment by sledgehammer999** [comment]
>as for a new flag, I basically imagine that being an option to just stop the torrent if check_resume fails for any reason, or something along those lines

Great.

>My understanding is that on unix systems it's possible, if not likely, that a path will exist even though the drive is not mounted. Perhaps it's limited to the root of the mount-point though, but still, it suggests that just checking to see if the path is there wouldn't be enough (because nothing can exist under that directory until the drive is mounted).

I have some experience with linux, but don't use it everyday. I haven't encountered this though. At least for automatically mounted drives. I see your point for fstab'ed ones. The mount point will exist regardless. Just the rest of the file-tree won't be available until the drive is mounted.
(I am be coming back to your original point)-> Check for existence of 1-2 files, which, based of the fastresume, should exist. But stat() would need to differentiate if it can't find it because of unmounted drive.

**Comment by sledgehammer999** [comment]
One last point: In case you introduce the add_params_flag, which on failure will pause the torrent -> then maybe you shouldn't even try to differentiate why it isn't available. What I mean is that this flag could mean "pause when files are not found/not match/etc and don't do a recheck yet". As a last resort at least in case you can't reliably detect missing drives.

**Comment by cas--** [comment]
This may/may not help but I thought I would mention that I recently committed a workaround for this issue in Deluge: [0ab7ebd0177f281e34a7](http://git.deluge-torrent.org/deluge/commit/?h=1.3-stable&id=0ab7ebd0177f281e34a706258c2c93cb22f4d833)

Obviously it will be great for the future if Deluge can rely on libtorrent taking care of this functionality. While I was coding this up I was pondering whether libtorrent should have a `set_error` function for a torrent for clients to set their own error, similar to my Deluge implementation?

**Comment by sledgehammer999** [comment]
I was re-reading the discussion. I want to add one small thing. @arvidn said:
>Can you think of anything to distinguish this case from the case where the destination files were just deleted?

Currently qbt handles both cases the same. And I don't see a reason to differentiate. User sees status "Missing files". If they deleted them themselves, they can either delete the torrent from the session or resume it (which will probably result in rechecking and download).

**Comment by arvidn** [comment]
does that mean distinguishing between resume failure because there are no files and failure because the files have been modified since the resume data was saved?
I'm envisioning a simple flag that will stop the torrent (in force-stopped checking mode) if the resume data fails. basically, the meaning of setting it is: "I'm sure the resume data is up-to-date, if it fails, I want to decide what to do rather than to recheck files".
does that make sense?
In the future (as indicated on the mailing list) I envision that the only failure of resume data checking will be one or more files actually missing, never just modified.

**Comment by sledgehammer999** [comment]
>I'm envisioning a simple flag that will stop the torrent (in force-stopped checking mode) if the resume data fails. basically, the meaning of setting it is: "I'm sure the resume data is up-to-date, if it fails, I want to decide what to do rather than to recheck files".

This will be perfect for my situation.

>In the future (as indicated on the mailing list) I envision that the only failure of resume data checking will be one or more files actually missing, never just modified.

I understand. But I don't know what will happen when the files will be indeed modified. You siad that it will just start downloading from the state that was saved in the resume data and possibly overwriting(and redownloading) valid data.
IMO, users will not like this. I could be wrong here. For this particular thing I tag you @glassez too. For reference, here is the mailing list [thread](http://sourceforge.net/p/libtorrent/mailman/message/34653954/) that arvidn is refering to.

**Comment by arvidn** [comment]
actually. while working on this, I realized the facilities to do this already exist (in master).

If the torrent is added in paused state (and not auto-managed), the fastresume should still be checked, but if it's rejected, the actual checking of files should not start (since it's paused). If the resume data fails verification, that alert is posted. In either case, the state_changed_alert is posted to indicate that it changed state into checking, downloading or seeding.

My new plan is to add tests to cover these cases.

I realize this is pretty far from the original topic of this thread. Am I forgetting something? Is this too impractical?

**Comment by arvidn** [comment]
@sledgehammer999 are you saying users actually rely on the ability to modify the files and then seed them? (expecting the files to be re-checked).

I suppose there may be inadvertent modifications like media players re-writing tags etc. But my understanding is that incorrectly re-checking the files when they don't need to be is a bigger problems for users. But perhaps the true problem is that users *believe* the files don't need to be rechecked, because they've been modified behind the users' back by some other application.

**Comment by sledgehammer999** [comment]
For the original question:

To do what you suggest means that we need to override the fastresume. And in turn somehow read ourselves the saved data for the other flags that don't belong to the `auto_manage` and `paused` state.
I thought of a way to force it to not download something before we get a chance to pause it: Pass the `flag_upload_mode` flag which seems to not be tied with `flag_override_resume_data`. I didn't test it and I don't know if it will work.

As for the mailing list issue:
I am be overthinking. IIRC one of the selling points of bittorrent when it was first introduced back then when we had really limited network speeds: "You can start downloading in X client and then continue using Y client without losing progress."
As you may point out nowadays this might be a feature that isn't used and you'll be right. If it is about what is more likely "the user to delete a file" or "the user modify the data by downloading or metadata rewriting". I would vote "the user deleting a file" is much more likely.
About verification: You could do hash check a piece when you're uploading it. Normally this shouldn't cause an performance degradation, right?

**Comment by arvidn** [comment]
presumably the fastresume data is incompatible between client X and Y, which means you'll always re-check the files.

**Comment by arvidn** [comment]
I'm not sure what you mean by your first paragraph. I may need to re-read the whole thread. But what I'm suggesting is: always add all torrents paused and non-auto-managed. That way, you will always get an opportunity to _not_ start it if the resume data fails. Or if it enters checking-file mode when you thought it shouldn't. If all is fine and the resume data was accepted, you just set auto_managed(true) and resume()

**Comment by sledgehammer999** [comment]
>But what I'm suggesting is: always add all torrents paused and non-auto-managed

These settigns are meaningless if I don't also pass `flag_override_resume_data`. However, that flag has other undesired effects (ie I don't want to override all saved settings)

**Comment by arvidn** [comment]
I see. because the paused flag may be changed.

**Comment by arvidn** [comment]
I wonder if there's a way to solve this that reduces complexity rather than increases it

**Comment by arvidn** [comment]
How about having a separate function to apply torrent state from resume data?

Basically, no torrent state other than relating to what's downloaded would be loaded when override resume data is set. It could be loaded by calling another function with the same resume data.

I assume the undesired effects of setting override resume data is that you then have to manually parse the resume data to know whether or not the torrent is supposed to be started or not (for instance, but also for all other state).

Actually, come to think of it. The most reasonable way of dealing with this is probably to move _all_ fields that exist in the resume data today into add_torrent_params, and make a separate function that parses resume data and fills in the add_torrent_params object. That way the client has full control and knowledge.

**Comment by sledgehammer999** [comment]
>The most reasonable way of dealing with this is probably to move all fields that exist in the resume data today into add_torrent_params, and make a separate function that parses resume data and fills in the add_torrent_params object. That way the client has full control and knowledge.

This makes more sense IMO.

**Comment by arvidn** [comment]
I've started this work on a branch called "read-resume-data". https://github.com/arvidn/libtorrent/tree/read-resume-data

**Comment by ruslanfedoseenko** [comment]
Will this changes be back portable to 1.1 or 1.0.x? Or master is quiet stable to use it?

**Comment by arvidn** [comment]
I don't intend to back-port this. It's an ABI altering change, and back-porting it would just postpone 1.1 even longer (it's been several years now)

**Comment by arvidn** [comment]
I also try to keep master stable. most of the time I believe it is

**Comment by arvidn** [comment]
Since the read_resume_data() branch has landed in master, I'll close this.
